compiler/src/main/scala/org/scalajs/nscplugin/GenJSExports.scala

```diff
@@ -71,189 +71,147 @@ trait GenJSExports[G <: Global with Singleton] extends SubComponent {
         .map(genJSClassDispatcher(classSym, _))
     }
 
-    def genConstructorExports(
-        classSym: Symbol): List[js.TopLevelMethodExportDef] = {
+    private sealed trait ExportKind
+
+    private object ExportKind {
+      case object Module extends ExportKind
+      case object JSClass extends ExportKind
+      case object Constructor extends ExportKind
+      case object Method extends ExportKind
+      case object Property extends ExportKind
+      case object Field extends ExportKind
+
+      def apply(sym: Symbol): ExportKind = {
+        if (isStaticModule(sym)) Module
+        else if (sym.isClass) JSClass
+        else if (sym.isConstructor) Constructor
+        else if (!sym.isMethod) Field
+        else if (jsInterop.isJSProperty(sym)) Property
+        else Method
+      }
+    }
 
-      val constructors = classSym.tpe.member(nme.CONSTRUCTOR).alternatives
+    private def checkSameKind(tups: Iterable[(jsInterop.ExportInfo, Symbol)]): Option[ExportKind] = {
+      assert(tups.nonEmpty, "must have at least one export")
 
-      // Generate exports from constructors and their annotations
-      val ctorExports = for {
-        ctor <- constructors
-        exp  <- jsInterop.registeredExportsOf(ctor)
-      } yield (exp, ctor)
+      val firstSym = tups.head._2
+      val overallKind = ExportKind(firstSym)
+      var bad = false
 
-      if (ctorExports.isEmpty) {
-        Nil
-      } else {
-        val exports = for {
-          (jsName, specs) <- ctorExports.groupBy(_._1.jsName) // group by exported name
-        } yield {
-          val ctors = specs.map(s => ExportedSymbol(s._2))
+      for ((info, sym) <- tups.tail) {
```

Comment:
`.tail` is not necessarily efficient on arbitrary `Iterable`s. Can you take a `List` instead? Otherwise, it might be better to iterate over `tups` rather than `tups.tail`.